test/extended/prometheus/statefulsets_stability.go

```diff
@@ -0,0 +1,96 @@
+package prometheus
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	g "github.com/onsi/ginkgo/v2"
+	o "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/upgrades"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// TODO: add other statefulsets
+var statefulsetsToCheck = []string{"prometheus-k8s"}
+
+// StatefulsetsStabilityAfterUpgradeTest tests that some statefulsets are not recreated
+// after the upgrade to avoid downtime.
+type StatefulsetsStabilityAfterUpgradeTest struct {
+	statefulsetsUID map[string]string
+}
+
+func (t *StatefulsetsStabilityAfterUpgradeTest) Name() string {
+	return "statefulset-stability-after-upgrade"
+}
+
+func (t *StatefulsetsStabilityAfterUpgradeTest) DisplayName() string {
+	return "[sig-instrumentation] Statefulsets are not recreated after upgrade"
+}
+
+func (t *StatefulsetsStabilityAfterUpgradeTest) Setup(ctx context.Context, f *e2e.Framework) {
+	defer g.GinkgoRecover()
+	g.By("getting the UID of the statefulsets before the upgrade")
+	t.statefulsetsUID = getStatefulsetsUID(ctx, f)
+}
+
+func (t *StatefulsetsStabilityAfterUpgradeTest) Test(
+	ctx context.Context,
+	f *e2e.Framework,
+	done <-chan struct{},
+	upgrade upgrades.UpgradeType,
+) {
+	<-done
+	defer g.GinkgoRecover()
+
+	g.By("statefulsets UID should stay the same after the upgrade")
+	o.Expect(assert.Equal(g.GinkgoT(), t.statefulsetsUID, getStatefulsetsUID(ctx, f))).
+		To(o.BeTrue())
+	o.Expect(assert.Equal(g.GinkgoT(), t.statefulsetsUID, map[string]string{"prometheus-k8s": "foo"})).
+		To(o.BeTrue())
+}
+
+func (t StatefulsetsStabilityAfterUpgradeTest) Teardown(_ context.Context, _ *e2e.Framework) {
+	// nothing
+}
+
+func (t *StatefulsetsStabilityAfterUpgradeTest) Skip(_ upgrades.UpgradeContext) bool {
+	return false
+}
+
+func getStatefulsetsUID(ctx context.Context, f *e2e.Framework) map[string]string {
+	openshiftMonitoringNs := "openshift-monitoring"
+	statefulsetsUID := make(map[string]string)
+	for _, statefulsetName := range statefulsetsToCheck {
+		err := wait.PollUntilContextTimeout(
+			ctx,
+			time.Second,
+			1*time.Minute,
+			true,
+			func(context.Context) (bool, error) {
+				statefulset, err := f.ClientSet.AppsV1().
+					StatefulSets(openshiftMonitoringNs).
+					Get(ctx, statefulsetName, metav1.GetOptions{})
+				if err != nil {
+					return false, err
```

Comment:
we need to save the error and return "false, nil" instead otherwise any Get error will stop the loop.